src/pymor/reductors/era.py

```diff
@@ -0,0 +1,237 @@
+    the Hankel matrix, i.e.
+
+    .. math::
+        \hat{h}_i=W_1^Th_iW_2,
+
+    where :math:`l_1\leq p` and :math:`l_2\leq m` are the number of left and right tangential
+    directions and :math:`W_1\in\mathbb{R}^{p\times l_1}` and :math:`W_2\in\mathbb{R}^{m\times l_2}`
+    are the left and right projectors, respectively. See :cite:`KG16`.
+
+    Attributes
+    ----------
+    data
+        |NumPy array| that contains the first :math:`n` Markov parameters of an LTI system.
+        Has to be one- or three-dimensional with either::
+
+            data.shape == (n,)
+
+        for scalar-valued Markov parameters or::
+
+            data.shape == (n, p, m)
+
+        for matrix-valued Markov parameters of dimension :math:`p\times m`, where
+        :math:`m` is the number of inputs and :math:`p` is the number of outputs of the system.
+    sampling_time
+        A number that denotes the sampling time of the system (in seconds).
+    force_stability
+        Whether the Markov parameters are zero-padded to double the length in order to enforce
+        Kung's stability assumption. See :cite:`K78`. Defaults to `True`.
+    feedthrough
+        (Optional) |Operator| or |Numpy array| of shape `(p, m)`. The zeroth Markov parameter that
+        defines the feedthrough of the realization. Defaults to `None`.
+    """
+
+    cache_region = 'memory'
+
+    def __init__(self, data, sampling_time, force_stability=True, feedthrough=None):
+        assert sampling_time >= 0
+        assert feedthrough is None or isinstance(feedthrough, np.ndarray) or isinstance(feedthrough, Operator)
+        assert np.isrealobj(data)
+        if data.ndim == 1:
+            data = data.reshape(-1, 1, 1)
+        assert data.ndim == 3
+        if isinstance(feedthrough, np.ndarray):
+            feedthrough = NumpyMatrixOperator(feedthrough)
+        if isinstance(feedthrough, Operator):
+            assert feedthrough.range.dim == data.shape[1]
+            assert feedthrough.source.dim == data.shape[2]
+        self.__auto_init(locals())
+
+    @cached
+    def _s1_W1(self):
+        self.logger.info('Computing output SVD ...')
+        W1, s1, _ = spla.svd(np.hstack(self.data), full_matrices=False)
+        return s1, W1
+
+    @cached
+    def _s2_W2(self):
+        self.logger.info('Computing input SVD ...')
+        _, s2, W2 = spla.svd(np.vstack(self.data), full_matrices=False)
+        return s2, W2.T
+
+    def _project_markov_parameters(self, num_left, num_right):
+        mats = [self.output_projector(num_left).T, self.data] if num_left else [self.data]
+        mats = [*mats, self.input_projector(num_right)] if num_right else mats
+        s1 = ('lp,', 'l') if num_left else ('', 'p')
+        s2 = (',mk', 'k') if num_right else ('', 'm')
+        self.logger.info('Projecting Markov parameters ...')
+        einstr = s1[0] + 'npm' + s2[0] + '->n' + s1[1] + s2[1]
+        return np.einsum(einstr, *mats, optimize='optimal')
```

Comment:
How does this compare to using `@`?

```python
result = self.data
if num_left:
    result = self.output_projector(num_left).T @ result
if num_right:
    result = result @ self.input_projector(num_left).T
return res
```

Reply:
I followed this approach initially but felt the need to improve it after doing some numerical tests. I have an example here where `era.data.shape = (256, 2, 11950)`. 

Running `era.reduce(256)` takes 14.6 seconds with about 1.3 seconds spent in `_project_markov_parameters` with `einsum`. Under the hood, `num_right` is set to `512`.

If I use the naive approach with `matmul`, `_project_markov_parameters` takes 14.9 seconds (29.5 total), so funnily exactly as long for the projection as the entire algorithm.

There are other cases with more conventionally shaped data, where the difference is not so pronounced. Also, in this example it was necessary to set `optimize='optimal'` to really get a noticeable speedup.

Although I don't have data that need two-sided projection yet, I believe that the speedup will be even greater when projecting both sides - similar to the argument for using `np.linalg.multi_dot` over `np.dot`. `multi_dot` does not support broadcasting which is why I ended up implementing it like it is now.

Reply:
Interesting. I also get that `einsum` is faster for one-sided projection.

```python
In [1]: import numpy as np

In [2]: A = np.random.randn(256, 2, 1000)

In [3]: V = np.random.randn(1000, 100)

In [4]: %timeit A @ V
15.5 ms ± 25.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [5]: %timeit np.einsum('npm,mk->npk', A, V, optimize='optimal')
863 µs ± 73.6 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

But, it looks like `einsum` is slower for two-sided projection.

```python
In [1]: import numpy as np

In [2]: A = np.random.randn(256, 1000, 1000)
 
In [3]: V = np.random.randn(1000, 100)

In [4]: W = np.random.randn(1000, 100)

In [5]: %timeit W.T @ A @ V
1.08 s ± 45.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [6]: %timeit np.einsum('lp,npm,mk->nlk', W.T, A, V, optimize='optimal')
2.61 s ± 166 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Reply:
Wow, that's an unexpected result! I don't know what `einsum` does under the hood, but `optimize='optimal'` definitely suggests that it should not be slower than `np.dot` (disregarding the time it takes to find the optimal path).

Reply:
I get that calling `einsum` twice in the second example is slightly faster than `@`:

```python

In [9]: %%timeit
   ...: AV = np.einsum('npm,mk->npk', A, V, optimize='optimal')
   ...: WTAV = np.einsum('lp,npk->nlk', W.T, AV, optimize='optimal')
   ...: 
   ...: 
990 ms ± 22.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Reply:
I just played around a bit more and it looks like the culprit might acutally be the left sided projection. I suspect it has to do with the memory layout. Does that give you an idea?

``` python
In [1]: import numpy as np

In [2]: A = np.random.randn(256, 1000, 1000)

In [3]: V = np.random.randn(1000,100)

In [4]: W = np.random.randn(1000,100)

In [5]: %timeit np.einsum('npm,mk->npk', A, V, optimize='optimal')
861 ms ± 19.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [6]: %timeit np.einsum('lp,npm->nlm', W.T, A, optimize='optimal')
1.86 s ± 59.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [7]: %timeit np.einsum('lp,npm,mk->nlk', W.T, A, V, optimize='optimal')
2.18 s ± 46.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [8]: %timeit np.einsum('npm,mk->npk', A.transpose(0,2,1), W, optimize='optimal')
1.64 s ± 78.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [9]: A.flags
Out[9]: 
  C_CONTIGUOUS : True
  F_CONTIGUOUS : False
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False

In [10]: A.transpose(0,2,1).flags
Out[10]: 
  C_CONTIGUOUS : False
  F_CONTIGUOUS : False
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
```

Reply:
> I get that calling `einsum` twice in the second example is slightly faster than `@`:
> 
> ```python
> In [9]: %%timeit
>    ...: AV = np.einsum('npm,mk->npk', A, V, optimize='optimal')
>    ...: WTAV = np.einsum('lp,npk->nlk', W.T, AV, optimize='optimal')
>    ...: 
>    ...: 
> 990 ms ± 22.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
> ```



> I get that calling `einsum` twice in the second example is slightly faster than `@`:
> 
> ```python
> In [9]: %%timeit
>    ...: AV = np.einsum('npm,mk->npk', A, V, optimize='optimal')
>    ...: WTAV = np.einsum('lp,npk->nlk', W.T, AV, optimize='optimal')
>    ...: 
>    ...: 
> 990 ms ± 22.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
> ```

``` python
In [11]: %timeit W.T @ A
958 ms ± 14.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [12]: %timeit np.einsum('lp,npm->nlm', W.T, A, optimize='optimal')
1.61 s ± 326 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [13]: %%timeit
    ...: AV = np.einsum('npm,mk->npk', A, V, optimize='optimal')
    ...: WTAV = np.einsum('lp,npk->nlk', W.T, A, optimize='optimal')
    ...: 
    ...: 
3 s ± 90.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [14]: %%timeit
    ...: AV = np.einsum('npm,mk->npk', A, V, optimize='optimal')
    ...: WTAV = W.T @ AV
    ...: 
    ...: 
1.07 s ± 17.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

A combination seems to be even faster. Maybe we should always do the right ones with einsum and left with matmul. I am still unsure about the order and if this can be generalized to other data shapes.

Reply:
@pmli If you agree, I would simply change to the latest suggestion.